### ISSUE TITLE
Dependency API Issues

### ISSUE THREAD

**Issue by seldridge** [freechipsproject/firrtl#1120]
**Type of issue**: API concern

**Impact**: API modification

**Development Phase**: request

## Summary

There are some concerns that the merged Dependency API (#1079) may be insufficient for all use cases, notably Aspects (https://github.com/freechipsproject/chisel3/pull/1077/) and repeated application of the a sequence of phases (https://github.com/ucb-bar/barstools/commit/787740).

I'm marking this as Milestone 1.2.0 so that either this is either blocking on 1.2.0 and this gets fixed or it's bumped back to 1.3.0 and we back out #1079.

Roughly, there's a couple of use cases where you would like to run the same `Transform` multiple times with different arguments/annotations. 

### Aspects

`$toBePopulated`

### Nested Group and Dedup

`As a Chisel user (driving towards a tape-out), I would like to be able to recursively group components into modules.`

Put differently, this means that a user wants to run `firrtl.transforms.GroupComponents` multiple times back-to-back. This also means that the user wants to run `firrtl.transforms.Dedup` between groupings. (Equivalently, this is back-to-back runs of `firrtl.transforms.GroupAndDedup`.) This is complicated by the fact that each run of `GroupComponents` needs to have a different set of `GroupAnnotation`s to get the correct ordering.

Currently, there is no way for the Dependency API to represent such an ordering because prerequisites and dependents are of type `Seq[Class[A]]` meaning *there exists no way to define dependencies between objects*.

## Proposed Solutions/Partial Solutions

Proposed ideas are summarized below. Note: these are sometimes overlapping and sometimes mutually exclusive.

### Dependencies Specified as Objects

Currently, the Dependency API specifies dependencies as classes. This could be converted to objects.

### DependencyManager Inputs are Objects

The Dependency API could stay as classes, but the inputs to a `DependencyManager`, "targets" and "current state", could be objects. 

There could be an optional guarantee that the ordering of the objects specified must be preserved.

### DependencyManager Supports Objects and Classes

Define targets as both objects and classes.

### Define a "Mapping" Annotation

Some of these ideas need a way of mapping specific annotations to specific transforms. This would require defining some function that splits/mangles/duplicates the `AnnotationSeq` so that the requested annotations are given to a specific instance of a transform.

### Non-Solution (Stacked Annotations)

Alternatively, this is working as intended and the Dependency API should stay as specifying dependencies as classes. This would then mean that for users who want object dependencies, they need to hide this inside a Dependency API-compliant class.

One non-solution a user could implement is to use a stack of annotations. Annotations can be packed into a "stacks" of `Seq[AnnotationSeq, Seq[AnnotationSeq]]`. This stack results in repeated application of a `Transform` with dependency resolution between applications until the stack is empty.

### 2019-07-10 

Participants: Adam, Schuyler

#### Differences of Aspect Ordering

Considering a design (`D`) and two aspects (`A1` and `A2`), does the ordering in which the aspects are applied matter? Or, is aspect "application" commutative? Further or: are the following equivalent?

```
(D + A1) + A2
(D + A2) + A1
```

If the aspect is an injecting aspect (adding some Chisel code), then no, due to last connect semantics. Consider an aspect that sets a signal to zero and an aspect that sets that same signal to one. Due to last connect semantics, the later applied aspect overwrites the earlier applied aspect.

#### Annotation Binding

Consider the following annotation sequence and sequence of transforms:

```scala
val annos: AnnotationSeq = Seq(Z(1), X, Z(2), Y)
val transforms: Seq[Transform] = Seq(TZ, TX, TZ, TY)
```

If a user wants to send `Z(1)` to the first `TZ` and the second `Z(2)` to the second `TZ`, how would you do this?

After some back and forth, the tentative conclusion is that this is not allowable. A user should have one application of `TZ` and an annotation that contains the **merged** version of `Z(1)` and `Z(2)`. 

Consider another example of two annotations that set the initialization of a register target to a specific value:
```scala
/* An annotation that will set the initial value of target to value */
case class SetTo(target: Target, value: Int)

/* Two annotations, the first sets foo to one, the second sets foo to zero */
val annos: AnnotationSeq = Seq(SetTo(foo, 1), SetTo(foo, 0)) 
```

Here, if one of these is passed to one resolving annotation and the other to a second instance, then the order in which the transforms runs determines the final value of the register `foo`.

If these two annotations are both emitted by Chisel, then there needs to be some way to resolve what the user intent was. @azidar proposed the concept of **merging** annotations. Here annotations define how they should be merged together and a `Phase` does this (similar to how annotations are updated based on their specified renaming rules). In effect, annotation merging is a means of coalescing multiple annotation input arguments to a transform down to one object.

This removes the need for the Dependency API to consider ordering multiple instances of the same class.

**Comment by seldridge** [comment]
After additional discussion on this, the (current) conclusion is that:

1. Ordering of transform instances is out of scope of the dependency manager.
2. If a transform writer wants to implement their *first transform* in such a way that the transform calls multiple instances of some other *second transform*, it's their job to both use a packed annotation that contains enough information for the *first transform* to know how to run instances of the *second transform.

This likely pushes towards a refactor for one multi-target annotation per transform as opposed to scattered single-target annotations.

The caveat here is that neither I nor @azidar were able to come up with a motivating use case that *required* being able to express instance orderings and could not be handled in some other way (e.g., packed annotations).

Closing this (and unblocking 1.2.0) pending ideas on further use cases.
